librato.go

```diff
@@ -29,29 +29,54 @@ type Reporter struct {
 	Percentiles     []float64              // percentiles to report on histogram metrics
 	TimerAttributes map[string]interface{} // units in which timers will be displayed
 	intervalSec     int64
+	closer          chan struct{}
 }
 
 func NewReporter(r metrics.Registry, d time.Duration, e string, t string, s string, p []float64, u time.Duration) *Reporter {
-	return &Reporter{e, t, s, d, r, p, translateTimerAttributes(u), int64(d / time.Second)}
+	return &Reporter{e, t, s, d, r, p, translateTimerAttributes(u), int64(d / time.Second), make(chan struct{})}
 }
 
 func Librato(r metrics.Registry, d time.Duration, e string, t string, s string, p []float64, u time.Duration) {
 	NewReporter(r, d, e, t, s, p, u).Run()
 }
 
+func (self *Reporter) Close() error {
+	close(self.closer)
+	return nil
+}
+
 func (self *Reporter) Run() {
 	ticker := time.Tick(self.Interval)
 	metricsApi := &LibratoClient{self.Email, self.Token}
-	for now := range ticker {
-		var metrics Batch
-		var err error
-		if metrics, err = self.BuildRequest(now, self.Registry); err != nil {
-			log.Printf("ERROR constructing librato request body %s", err)
-			continue
-		}
-		if err := metricsApi.PostMetrics(metrics); err != nil {
-			log.Printf("ERROR sending metrics to librato %s", err)
-			continue
+
+	for {
+		select {
+		case now := <-ticker:
+			var metrics Batch
+			var err error
+			if metrics, err = self.BuildRequest(now, self.Registry); err != nil {
+				log.Printf("ERROR constructing librato request body %s", err)
+				break
+			}
+			if err := metricsApi.PostMetrics(metrics); err != nil {
+				log.Printf("ERROR sending metrics to librato %s", err)
+				break
+			}
+
+		case <-self.closer:
+			now := time.Now()
+			var metrics Batch
+			var err error
+			if metrics, err = self.BuildRequest(now, self.Registry); err != nil {
+				log.Printf("ERROR constructing librato request body %s", err)
+				return
+			}
+			if err := metricsApi.PostMetrics(metrics); err != nil {
+				log.Printf("ERROR sending metrics to librato %s", err)
+				return
+			}
+			log.Printf("go-metrics-librato: close received, cleaned up in %dms", time.Since(now)/time.Millisecond)
+			return
 		}
 	}
```

Comment:
For the love of all things holy, refactor this into a func.

Can easily return `err` and break in the first case, and return `err` and return in the second.